### PR TITLE
Handle length-prefixed messages of length 0

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto2WireProtocCompatibilityTests.kt [hunk 1]
```diff
@@ -156,6 +161,40 @@ class Proto2WireProtocCompatibilityTests {
...
+    val googleMessage = RepeatedEnum.QuiltContainer.newBuilder()
+      .setQuilt(
+        RepeatedEnum.Quilt.newBuilder()
+          .addAllFringe(listOf(RepeatedEnum.QuiltColor.GREEN))
+          .setCozy(true)
+          .build()
+      ).build()
+
+    val wireMessageDecodedFromGoogleMessage =
+      QuiltContainer.ADAPTER.decode(encodedByTsProto)
```
Comment (oldergod) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/Proto2WireProtocCompatibilityTests.kt:190]: That's interesting. This line fails on CI with
```
java.io.IOException: Expected to end at 7 but was 8
        at com.squareup.wire.ProtoReader.afterPackableScalar(ProtoReader.kt:385)
        at com.squareup.wire.ProtoReader.readVarint32(ProtoReader.kt:277)
        at com.squareup.wire.EnumAdapter.decode(EnumAdapter.kt:72)
        at com.squareup.wire.EnumAdapter.decode(EnumAdapter.kt:25)
        at squareup.proto2.kotlin.interop.Quilt$Companion$ADAPTER$1.decode(Quilt.kt:205)
        at squareup.proto2.kotlin.interop.Quilt$Companion$ADAPTER$1.decode(Quilt.kt:159)
        at squareup.proto2.kotlin.interop.QuiltContainer$Companion$ADAPTER$1.decode(QuiltContainer.kt:110)
        at squareup.proto2.kotlin.interop.QuiltContainer$Companion$ADAPTER$1.decode(QuiltContainer.kt:82)
```

But protogram which is still on Wire 3.1.0 can deserialize it https://mattprecious.github.io/protogram/?12091201041a0022003801

image

Did we break in the mean time?

File: wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/alltypes/AllTypes.kt [hunk 2]
```diff
@@ -2686,73 +2698,73 @@ public class AllTypes(
...
                 array_sfixed32 = IntArrayList.forDecoding(reader.nextFieldMinLengthInBytes(), 4)
               }
-              array_sfixed32!!.add(com.squareup.wire.`internal`.decodePrimitive_sfixed32(reader))
+              array_sfixed32!!.add(decodePrimitive_sfixed32(reader))
             }
             606 -> {
               if (array_int64 == null) {
                 array_int64 = LongArrayList.forDecoding(reader.nextFieldMinLengthInBytes(), 1)
               }
-              array_int64!!.add(com.squareup.wire.`internal`.decodePrimitive_int64(reader))
```
Comment (swankjesse) [inline wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/alltypes/AllTypes.kt:2719]: These becoming not-fully-qualified is alarming

Comment (swankjesse) [reply wire-tests/src/commonTest/proto-kotlin/com/squareup/wire/protos/kotlin/alltypes/AllTypes.kt:2719]: .... ahhhh it’s cause I stopped doing unnecessary toString() on 'em. Never mind then!

File: wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/alltypes/AllTypes.kt [hunk 3]
```diff
@@ -4585,10 +4597,10 @@ public class AllTypes(
                   .toInt()
                 ext_pack_double = ArrayList(initialCapacity)
               }
-              ext_pack_double!!.add(com.squareup.wire.ProtoAdapter.DOUBLE.decode(reader))
+              ext_pack_double!!.add(ProtoAdapter.DOUBLE.decode(reader))
             }
             1216 -> try {
-              ext_pack_nested_enum.add(NestedEnum.ADAPTER.decode(reader))
+              NestedEnum.ADAPTER.tryDecode(reader, ext_pack_nested_enum)
```
Comment (swankjesse) [inline wire-tests/src/jvmKotlinInteropTest/proto-kotlin/com/squareup/wire/protos/kotlin/alltypes/AllTypes.kt:4603]: This is the good stuff right here.

**Review by oldergod — APPROVED** [review]

File: wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoAdapter.kt [hunk 4]
```diff
@@ -144,6 +144,13 @@ expect abstract class ProtoAdapter<E>(
   @Throws(IOException::class)
   fun decode(source: BufferedSource): E
 
+  /**
+   * Reads a value and appends it to [destination] if this has data available. Otherwise, does
+   * nothing.
```
[the remark this answers is not in the document]
Comment (swankjesse) [reply wire-runtime/src/commonMain/kotlin/com/squareup/wire/ProtoAdapter.kt]: Good catch.

**Pull request merged**